Expose filename for tensor/storages created with `from_file` and document torch.from_file

mikaylagawarecki opened the pull request:
Fixes https://github.com/pytorch/pytorch/issues/37439

Also threads through filename so it is accessible via `t.storage().filename`

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #111856
* __->__ #111688



cc @albanD

mikaylagawarecki committed:
Document torch.from_file and fix UntypedStorage.from_file docs

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Document torch.from_file and fix UntypedStorage.from_file docs"




[ghstack-poisoned]

mikaylagawarecki commented on torch/_torch_docs.py:
```diff
@@ -4489,6 +4489,42 @@ def merge_dicts(*dicts):
+add_docstr(
+    torch.from_file,
+    """
+from_file(filename, shared=None, size=0, *, dtype=None, layout=torch.strided, device=None) -> Tensor
```
This technically has `pin_memory` as an argument but I'm avoiding documenting it since `pin_memory` will always fail if set

mikaylagawarecki committed:
Update on "Document torch.from_file and fix UntypedStorage.from_file docs"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Document torch.from_file and fix UntypedStorage.from_file docs"

Fixes https://github.com/pytorch/pytorch/issues/37439




cc albanD

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Document torch.from_file and fix UntypedStorage.from_file docs"

Fixes https://github.com/pytorch/pytorch/issues/37439




cc albanD

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Document torch.from_file and fix UntypedStorage.from_file docs"

Fixes https://github.com/pytorch/pytorch/issues/37439




cc albanD

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Document torch.from_file and fix UntypedStorage.from_file docs"

Fixes https://github.com/pytorch/pytorch/issues/37439




cc albanD

[ghstack-poisoned]

albanD commented on test/test_tensor_creation_ops.py:
```diff
@@ -3501,6 +3502,16 @@ def test_randperm_device_compatibility(self, device):
+        with tempfile.NamedTemporaryFile() as f:
+            t.numpy().tofile(f)
+            t_mapped = torch.from_file(f.name, shared=False, size=t.numel(), dtype=dtype)
+            self.assertTrue(t_mapped.untyped_storage().filename == f.name)
```
nit: you can also check that t and t_mapped are equal

albanD commented on torch/utils/_tensor_creation.py:
```diff
@@ -0,0 +1,62 @@
+        Only CPU tensors can be mapped to files.
+
+    .. note::
+        For now, tensors with storages backed by a memory-mapped file cannot be copied to pinned memory.
```
nit: they can be copied to pinned memory right? They cannot be created in pinned memory here?

albanD reviewed (commented):
Thanks!
I think you're missing the documentation for the `filename` field on UntypedStorage?

mikaylagawarecki committed:
Update on "Document torch.from_file and fix UntypedStorage.from_file docs"

Fixes https://github.com/pytorch/pytorch/issues/37439




cc albanD

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Document torch.from_file and fix UntypedStorage.from_file docs"

Fixes https://github.com/pytorch/pytorch/issues/37439




cc albanD

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Document torch.from_file and fix UntypedStorage.from_file docs"

Fixes https://github.com/pytorch/pytorch/issues/37439




cc albanD

[ghstack-poisoned]

albanD commented on torch/storage.py:
```diff
@@ -329,6 +332,15 @@ def is_cuda(self):
+    def filename(self) -> _Optional[str]:
+        """Returns the file name associated with this storage if the storage was memory mapped from a file.
+           or ``None`` if the storage was not created by memory mapping a file."""
+        if self._filename is None:
```
This is a bit confusing.
I think you should either:
- Have a commend here explaining what None and -1 means for `._filename()`
- Always call `_get_filename()` as it is a cheap function anyways.

mikaylagawarecki replied:
Did the second!

albanD reviewed (approved):
nit on the attribute but SGTM otherwise!

mikaylagawarecki committed:
Update on "Document torch.from_file and fix UntypedStorage.from_file docs"

Fixes https://github.com/pytorch/pytorch/issues/37439




cc albanD

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Document torch.from_file and fix UntypedStorage.from_file docs"

Fixes https://github.com/pytorch/pytorch/issues/37439

Also threads through filename so it is accessible via `t.storage().filename()`




cc albanD

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Document torch.from_file and fix UntypedStorage.from_file docs"

Fixes https://github.com/pytorch/pytorch/issues/37439

Also threads through filename so it is accessible via `t.storage().filename()`




cc albanD

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Document torch.from_file and fix UntypedStorage.from_file docs"

Fixes https://github.com/pytorch/pytorch/issues/37439

Also threads through filename so it is accessible via `t.storage().filename`




cc albanD

[ghstack-poisoned]

mikaylagawarecki commented:
@pytorchbot merge
